c++/src/kj/async-inl.h

```diff
@@ -1470,4 +1470,290 @@ PromiseForResult<Func, void> Executor::executeAsync(Func&& func) const {
+  auto final_suspend() noexcept { return stdcoro::suspend_never(); }
+  // These adjust the suspension behavior of coroutines immediately upon initiation, and immediately
+  // after completion.
+  //
+  // The initial suspension point could allow us to defer the initial execution of a coroutine with
+  // evalLater(), for instance.
+  //
+  // The final suspension point could be useful to delay deallocation of the coroutine frame to
+  // match the lifetime of the enclosing promise.
+
+  void unhandled_exception() {
+    // Pretty self-explanatory, we just propagate the exception to the promise which owns us. Note
+    // that all unhandled exceptions end up here, not just ones after the first co_await.
+    adapter->fulfiller.rejectIfThrows([] { throw; });
+  }
+
+  template <typename U>
+  class Awaiter;
+
+  template <typename U>
+  Awaiter<U> await_transform(kj::Promise<U>& promise) { return Awaiter<U>(kj::mv(promise)); }
+  template <typename U>
+  Awaiter<U> await_transform(kj::Promise<U>&& promise) { return Awaiter<U>(kj::mv(promise)); }
+  // Called when someone writes `co_await promise`, where `promise` is a kj::Promise<U>. We return
+  // an Awaiter<U>, which implements coroutine suspension and resumption in terms of the KJ async
+  // event system.
+  //
+  // There is another hook we could implement: an `operator co_await()` free function. However, a
+  // free function would be unaware of the type of the enclosing coroutine. Since Awaiter<U> is a
+  // member class template of Coroutine<T>, it is able to implement an
+  // `await_suspend(Coroutine<T>::Handle)` override, providing it type-safe access to our enclosing
+  // coroutine's promise adapter. An `operator co_await()` free function would have to implement
+  // a type-erased `await_suspend(stdcoro::coroutine_handle<void>)` override, and implement
+  // suspension and resumption in terms of .then(). Yuck!
+
+private:
+  friend class CoroutineBase<Coroutine<T>, T>;
+
+  class Adapter;
+  Adapter* adapter = nullptr;
+  // Set by Adapter's ctor in `get_return_object()`. Nulled out again by Adapter's dtor, if the
+  // promise returned from `get_return_object()` is destroyed while this coroutine is suspended.
+};
+
+template <typename Self, typename T>
+class CoroutineBase {
+public:
+  void return_value(T value) {
+    static_cast<Self*>(this)->adapter->fulfiller.fulfill(kj::mv(value));
+  }
+};
+template <typename Self>
+class CoroutineBase<Self, void> {
+public:
+  void return_void() {
+    static_cast<Self*>(this)->adapter->fulfiller.fulfill();
+  }
+};
+// The Coroutines spec has no `_::FixVoid<T>` equivalent to unify valueful and valueless co_return
+// statements, and programs are ill-formed if the coroutine implementation object (Coroutine<T>) has
+// both a `return_value()` and `return_void()`. No amount of EnableIffery can get around it, so
+// these return_* functions live in a CRTP mixin.
+
+template <typename T>
+class Coroutine<T>::Adapter: private Event {
```

Comment:
Instead of using promise adapters, it may make sense to implement `PromiseNode` directly. All `newAdaptedPromise()` really does is wrap your class in an `AdapterPromiseNode` which is just some boilerplate to make it a friendlier public interface.

In fact, I think then you could possibly even make `Coroutine<T>` implement `PromiseNode` directly and avoid having to allocate the adapter separately.

Reply:
Converted to PromiseNode: https://github.com/capnproto/capnproto/pull/1104/commits/9428a2532df76491831ed96af54b18b577e88813
Made Coroutine<T> the PromiseNode: https://github.com/capnproto/capnproto/pull/1104/commits/99c2cd5c86b4d4660cc2133f2c8c9a701bcc4264